src/files.ml

```diff
@@ -3,6 +3,211 @@
 open Extra
 open Console
 
+(** Logging function for evaluation. *)
+let log_file = new_logger 'f' "file" "file system"
+let log_file = log_file.logger
+
+(** Representation of module paths and related operations. *)
+module Path =
+  struct
+    (** Representation of a module path (roughly, a file path). *)
+    type module_path = string list
+
+    (** Short synonym of [module_path]. *)
+    type t = module_path
+
+    (** [compare] is a standard comparing function for module paths. *)
+    let compare : t -> t -> int = Pervasives.compare
+
+    (** [pp oc mp] prints [mp] to channel [oc]. *)
+    let pp : module_path pp = fun oc mp ->
+      Format.pp_print_string oc (String.concat "." mp)
+  end
+
+(** Functional maps with module paths as keys. *)
+module PathMap = Map.Make(Path)
+
+(** Synonym of [string] for file paths. *)
+type file_path = string
+
+(** Representation of the mapping from module paths to files. *)
+module ModMap :
+  sig
+    (** Module path mapping. *)
+    type t
+
+    (** [empty] is an empty module path mapping. *)
+    val empty : t
+
+    (** Exception raised if an attempt is made to map an already mapped module
+        (including the root). *)
+    exception Already_mapped
+
+    (** [set_root path m] sets the library root of [m] to be [path]. If it was
+        already set, the the exception [Already_mapped] is raised. *)
+    val set_root : file_path -> t -> t
+
+    (** [add mp path m] extends the mapping [m] by associating the module path
+        [mp] to the file path [path]. In the case where [mp] is already mapped
+        in [m], the exception [Already_mapped] is raised. *)
+    val add : Path.t -> file_path -> t -> t
+
+    (** Exception raised if an attempt is made to use the [get] function prior
+        to the root being set (using [set_root]). *)
+    exception Root_not_set
+
+    (** [get mp m] obtains the file path corresponding to the module path [mp]
+        in [m] (with no particular extension). The exception [Root_not_set] is
+        raised if the root of [m] was not previously set with [set_root]. *)
+    val get : Path.t -> t -> file_path
+
+    (** [iter fn m] calls function [fn] on every binding stored in [m]. *)
+    val iter : (Path.t -> file_path -> unit) -> t -> unit
+  end =
+  struct
+    type t = Node of file_path option * t StrMap.t
+
+    let empty = Node(None, StrMap.empty)
+
+    exception Already_mapped
+
+    let set_root path (Node(po, m)) =
+      match po with
+      | None    -> Node(Some(path), m)
+      | Some(_) -> raise Already_mapped
+
+    let rec singleton ks path =
```

Comment:
What's the type of this function and what is it doing?

Reply:
The local module has a signature.

Reply:
And this function is not part of the interface (it is only an auxiliary function).